exercises/robot-simulator/example.rs

```diff
@@ -0,0 +1,101 @@
+#[derive(PartialEq, Debug, Copy, Clone)]
+pub enum Direction {
+    North,
+    East,
+    South,
+    West,
+}
+
+impl Direction {
+    pub fn previous_clockwise(&self) -> Self {
+        match self {
+            &Direction::North => Direction::West,
+            &Direction::East => Direction::North,
+            &Direction::South => Direction::East,
+            &Direction::West => Direction::South,
+        }
+    }
+
+    pub fn next_clockwise(&self) -> Self {
+        match self {
+            &Direction::North => Direction::East,
+            &Direction::East => Direction::South,
+            &Direction::South => Direction::West,
+            &Direction::West => Direction::North,
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+struct Position {
+    x: i32,
+    y: i32,
+}
+
+impl Position {
+    fn new(x: i32, y: i32) -> Self {
+        Position { x: x, y: y }
+    }
+
+    fn advance(&self, direction: &Direction) -> Self {
+        match direction {
+            &Direction::North => Self::new(self.x, self.y + 1),
+            &Direction::South => Self::new(self.x, self.y - 1),
+            &Direction::East => Self::new(self.x + 1, self.y),
+            &Direction::West => Self::new(self.x - 1, self.y),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Robot {
+    position: Position,
+    direction: Direction,
+}
+
+impl Robot {
+    pub fn new(x: i32, y: i32, d: &Direction) -> Self {
+        Robot::build(Position::new(x, y), d.clone())
+    }
+
+    fn build(position: Position, direction: Direction) -> Self {
+        Robot {
+            position: position,
+            direction: direction,
+        }
+    }
+
+    pub fn turn_right(&self) -> Self {
+        Self::build(self.position, self.direction.next_clockwise())
+    }
+
+    pub fn turn_left(&self) -> Self {
+        Self::build(self.position, self.direction.previous_clockwise())
+    }
+
+    pub fn advance(&self) -> Self {
+        Self::build(self.position.advance(&self.direction), self.direction)
```

Comment:
This is personal preference but I find these one liners rather noisy and think separating it out is clearer to see what's happening:
```rust
let mut next = self.clone();
next.position = next.position.advance(self.direction);
next
```

Or to pattern match it into components and rebuild:
```rust
let Robot { mut position, direction } = self.clone();
position = position.advance(direction);
Self::build(position, direction)
```

Or with functional update syntax:
```rust
Robot {
    position: self.position.advance(&self.direction),
    // makes it more clear that only the position property is changing
    .. self.clone() 
}
```